### PR TITLE
Using FetchContent_Declare to automatically setup OpenFST at configure time.

### PR DESCRIPTION
**Pull request by vzxxbacq**
fix https://github.com/kaldi-asr/kaldi/issues/4529

### CODE REVIEW

**Comment by kkm000** [comment]
Looks like a definite improvement, thanks! Does it build binaries as well? And did you test it on Linux, too?

Some recipes require FAR extensions of OpenFST (the configure switches in our tools/Makefile aren't exactly correct, I suspect). But this is not a show-stopper, we can take care of that next.

File: cmake/third_party/openfst.cmake [hunk 1]
```diff
@@ -1,14 +1,25 @@
...
-    BUILD_COMMAND ""
-    INSTALL_COMMAND ""
-    TEST_COMMAND ""
+cmake_minimum_required(VERSION 3.14)
+include(FetchContent)
+
+FetchContent_Declare(
+        openfst
+        GIT_REPOSITORY  https://github.com/kkm000/openfst
+        GIT_TAG         338225416178ac36b8002d70387f5556e44c8d05 # tag win/1.7.2.1
```
Comment (danpovey) [inline cmake/third_party/openfst.cmake:7]: tag win/1.7.2.1... does this mean it's specifically for Windows?

Comment (vzxxbacq) [reply cmake/third_party/openfst.cmake:7]: the comment , I test on ubuntu and WSL(arch)

> Looks like a definite improvement, thanks! Does it build binaries as well? And did you test it on Linux, too?
> 
> Some recipes require FAR extensions of OpenFST (the configure switches in our tools/Makefile aren't exactly correct, I suspect). But this is not a show-stopper, we can take care of that next.

I have built the project on ubuntu and WSL (arch) using ninja build. The configure switches depend on the CMakeLists.txt in the OpenFST project.

**Comment by vzxxbacq** [comment]
@kkm000  
I have built the project on ubuntu and WSL (arch) using ninja build. The configure switches depend on the CMakeLists.txt in the OpenFST project.

File: cmake/third_party/openfst.cmake (hunk 1 above)
Comment (vzxxbacq) [reply cmake/third_party/openfst.cmake:7]: The tag is the newest commit of the winport branch not specifically for Windows, I forgot to remove the old comment, sorry.

**Comment by danpovey** [comment]
I'm OK to merge it as is if it would work.

**Comment by kkm000** [comment]
> The configure switches depend on the CMakeLists.txt in the OpenFST project.

Yup, and that is kind of the gray areas of CMake. There is no generally accepted best practices for controlling the subproject's publicly exposed controls. The "strong-arm" approach is to set all (or, at the least, all required) exposed variables before including the subdirectory. This may be confusing, because project's cached options that you see in `ccmake` or the GUI are not those which are applied during build, and changes that the user applies to the cache have no effect. Another is to test if the variables are cached, and seed them only if they aren't. Sometimes, a combination of the two is necessary. 

But I digress here. I just want to be sure we build what is necessary for Kaldi _by default_. In other words, if the user just runs `cmake -B build_dir && cmake --build build_dir`. For OpenFST, this means FAR extensions and all corresponding binaries. We just need to make sure these are build in thise very default case. Let's commit this change so it does not blow up, and then look into this requirement and make sure it's satisfied.

For Kaldi proper, only the base libfst is required as a dependency; base FST binaries are obviously needed for all egs, and some call for more than the basic set. I think that the only extension required is FAR (binaries, not libraries: if Kaldi is built without binaries—there's a CMake option for that, and I think it's a good idea to have this option—then no extensions are needed).

**Pull request merged**